src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs

```diff
@@ -0,0 +1,853 @@
+IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: M[]) (Syntax: 'new[] { new M() }')
+  Dimension Sizes(1):
+      ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: 'new[] { new M() }')
+  Initializer: IArrayInitializer (1 elements) (OperationKind.ArrayInitializer) (Syntax: '{ new M() }')
+      Element Values(1):
+          IObjectCreationExpression (Constructor: M..ctor()) (OperationKind.ObjectCreationExpression, Type: M) (Syntax: 'new M()')
+            Arguments(0)
+            Initializer: null
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<ImplicitArrayCreationExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact, WorkItem(17596, "https://github.com/dotnet/roslyn/issues/17596")]
+        public void ArrayCreationWithInitializerErrorCase_ImplicitlyTypedWithoutInitializerAndDimension()
+        {
+            string source = @"
+class C
+{
+    public void F(int dimension)
+    {
+        var x = /*<bind>*/new[]/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: ?[], IsInvalid) (Syntax: 'new[]/*</bind>*/')
+  Dimension Sizes(1):
+      ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 0, IsInvalid) (Syntax: 'new[]/*</bind>*/')
+  Initializer: IArrayInitializer (0 elements) (OperationKind.ArrayInitializer, IsInvalid) (Syntax: '')
+      Element Values(0)
+";
+            var expectedDiagnostics = new DiagnosticDescription[] {
+                // CS1514: { expected
+                //         var x = /*<bind>*/new[]/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_LbraceExpected, ";").WithLocation(6, 43),
+                // CS1513: } expected
+                //         var x = /*<bind>*/new[]/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_RbraceExpected, ";").WithLocation(6, 43),
+                // CS0826: No best type found for implicitly-typed array
+                //         var x = /*<bind>*/new[]/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_ImplicitlyTypedArrayNoBestType, "new[]/*</bind>*/").WithLocation(6, 27)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<ImplicitArrayCreationExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact, WorkItem(17596, "https://github.com/dotnet/roslyn/issues/17596")]
+        public void ArrayCreationWithInitializerErrorCase_ImplicitlyTypedWithoutInitializer()
+        {
+            string source = @"
+class C
+{
+    public void F(int dimension)
+    {
+        var x = /*<bind>*/new[2]/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Int32[], IsInvalid) (Syntax: 'new[2]/*</bind>*/')
```

Comment:
>System.Int32[] [](start = 71, length = 14)

It feels strange that we decided the type the array is int[]. Please open an issue to investigate.

Reply:
https://github.com/dotnet/roslyn/issues/22315